index.bs

```diff
@@ -783,6 +917,63 @@ steps:
 
 # Accessing native filesystem # {#native-filesystem}
 
+## Permissions ## {#native-file-system-permissions}
+
+<div algorithm>
+The <dfn>native file system query permission steps</dfn> for an [=/entry=] |entry|
+given a {{FileSystemHandlePermissionDescriptor}} |descriptor| are:
+
+1. If |descriptor|.{{FileSystemHandlePermissionDescriptor/writable}} is true:
+   1. Let |readState| be the result of running the [=query permission steps=] for |entry|
+      given «[&nbsp;"{{FileSystemHandlePermissionDescriptor/writable}}" → `false`&nbsp;]».
+   1. If |readState| is not {{PermissionState/"granted"}}, return |readState|.
+1. Let |settings| be the [=current settings object=].
+1. If there was a previous invocation of this algorithm for the same |entry|, |descriptor|
+   and |settings|, returning |previousResult|, and the UA has not received
+   [=new information about the user's intent=] since that invocation, return |previousResult|.
+1. Return whichever of the following options most accurately reflects the user's
+   intent for the calling algorithm:
+   <dl class="switch">
+     : succeed without prompting the user
+     :: {{PermissionState/"granted"}}
+
+     : show the user a prompt to decide whether to succeed
+     :: {{PermissionState/"prompt"}}
+
+     : fail without prompting the user
+     :: {{PermissionState/"denied"}}
+   </dl>
+
+</div>
+
+<div algorithm>
+The <dfn>native file system request permission steps</dfn> for an [=/entry=] |entry|
+given a {{FileSystemHandlePermissionDescriptor}} |descriptor| are:
+
+1. Let |status| be the result of running the [=query permission steps=] for |entry| given |descriptor|.
+1. If |status| is not {{PermissionState/"prompt"}} return.
+1. If the [=current global object=] is not a {{Window}},
+   return.
+
+   Issue: Should this throw instead of return silently?
+
+1. If the [=current global object=] does not have [=transient activation=], throw a {{NotAllowedError}}.
+1. Let |environment| be the [=current settings object=].
+1. If |environment|'s [=environment settings object/origin=] is not [=same origin=] with |environment|'s [=top-level origin=],
+   throw a {{NotAllowedError}}.
+1. Display a prompt to the user requesting access as described by |descriptor| to |entry|.
```

Comment:
I suspect this could be even more abstract, i.e. allowing a UA that never shows a prompt but always rejects, or infers an answer through other means.

I'd follow whatever other specs do. (I'm too lazy to check...)

Reply:
This sort of mostly matches what the permissions API uses for its request permission algorithm. I think the "never show a prompt" case is largely captured by the query permission steps. I.e. a UA can use whatever it wants to figure out what "most accurately reflects the user's intent", and if what most accurately reflects is in fact "prompt", this algorithm probably should in fact display a prompt (and if not, we won't even get to this step, since we'll just early abort this algorithm).